sdks/python/apache_beam/transforms/cy_combiners.py

```diff
@@ -405,3 +408,22 @@ def increment_bucket(self, bucket_index):
 
 class DistributionCounterFn(AccumulatorCombineFn):
   _accumulator_type = DistributionAccumulator
+
+
+class DistributionMetadataAccumulator(object):
+  def __init__(self):
+    self.distribution_metadata = []
+
+  def add_input(self, element):
+    self.distribution_metadata.append(element)
```

Comment:
<!--new_thread; commit:c89749a0d77350dfbd2686db975aa087bafb8f1b; resolved:0-->
It looks like to me (looking at this change and the internal Dataflow worker change) that the reason we need the `DistributionMetadataAccumulator` / `DistributionMetadataCounterFn` classes is that `DistributionCounterFn` is too slow, so here, we add everything to a list, and we do processing for the list later.  Is my understanding correct?

If so, I don't think using this instead of a `DistributionCounter` counter directly changes much--the processing will just get done in the worker API client thread, right, where we create another DistributionCounter called `output_counter` and individually call `output_counter.update(value)` for each value we push onto the list here?  In such a case, we would be hiding the overall overhead of this counter from the processing here, but we don't get rid of it--we just move it to the progress reporting thread?

Reply:
Yes, it is. I think, moving distribution calculation to report_progress thread will reduce some overhead overall since report_progress thread doesn't block execution thread, comparing to all distribution calculated in runner code path.

Reply:
This is concerning to me because this overhead won't show up in benchmarks of the execution path here, but will be the same in practice--we are just moving it to another thread.  Python execution is essentially single-core-bound because of the Global Interpreter Lock (GIL; http://www.dabeaz.com/python/UnderstandingGIL.pdf).  From the internal Dataflow change, I see that what is being done is equivalent to using a DistributionCounter directly here.  Perhaps we should focus on optimizing the DistributionCounter code path so that this can be used directly here.

CC: @robertwb

Reply:
I have one PR which related to DistributionCounter optimization: https://github.com/apache/beam/pull/4928. I'm not sure whether there are other ways to make it better. Do you have some ideas on this?

Reply:
As @charlesccychen suggested, I ran map_fn_microbenchmark with DistributionCounter and DistributionMetadataCounter enabled to see whether it's reasonable to update DistributionCounter directly during execution. Results as below:
```
Experimental flag disabled:
     1 element  0.882056 sec
  1001 elements 1.06062 sec
  2001 elements 0.933104 sec
  3001 elements 1.1544 sec
  4001 elements 1.22668 sec
  5001 elements 1.24582 sec
  6001 elements 1.52356 sec
  7001 elements 1.50509 sec
  8001 elements 1.87628 sec
  9001 elements 1.845 sec
Fixed cost   0.824348012299
Per-element  1.11289043138e-06
R^2          0.919787959108
```
```
DistributionMetadataCounter enabled:
     1 element  0.903107 sec
  1001 elements 1.06191 sec
  2001 elements 1.1449 sec
  3001 elements 1.31494 sec
  4001 elements 1.3732 sec
  5001 elements 1.42241 sec
  6001 elements 1.83665 sec
  7001 elements 1.84627 sec
  8001 elements 1.9221 sec
  9001 elements 2.1069 sec
Fixed cost   0.893742784382
Per-element  1.33191660679e-06
R^2          0.969293745853
```
```
DIstributionCounter enabled with optimization(https://github.com/apache/beam/pull/4928)
Distribution enable:
     1 element  0.933321 sec
  1001 elements 1.02975 sec
  2001 elements 1.12517 sec
  3001 elements 1.45851 sec
  4001 elements 1.46911 sec
  5001 elements 1.64176 sec
  6001 elements 1.8858 sec
  7001 elements 2.03899 sec
  8001 elements 2.17605 sec
  9001 elements 2.33426 sec
Fixed cost   0.882123873349
Per-element  1.61552530346e-06
R^2          0.988375238657

```
CC: @robertwb, @pabloem

Reply:
Thanks!  @robertwb, can you look at this?  To my understanding, the `DistributionMetadataCounter` and `DistributionCounter` cases above should be equivalent--the `DistributionMetadataCounter` results hide the cost of doing the distribution update in the progress reporting thread, which isn't included in the benchmark.  We should therefore use the `DistributionCounter` numbers in assessing the cost of these changes.

Reply:
In non-IO-bound cases, this would contribute in the same way to the total runtime, so putting it in the progress reporting thread is not free.  I will let @robertwb make the decision on the acceptable cost threshold here.

Reply:
While moving computation to the progress reporting thread is not free, there are significant advantages, as the progress reporting thread is called at a much lower frequency (seconds vs. microseconds). This is assuming, of course, that computing a distribution can be amortized across several elements in a manner cheaper than handling each one individually (batched or sequentially). 

It may be an open question whether distributions of the output counts of every single element for every single DoFn can be made cheap enough to enable everywhere; certainly the 20% overhead here is too high. To stay under 5% (using half of a 10% "monitoring budget" for this single feature) we need a distribution structure that can be updated in 50ns. This is certainly possible in C (e.g. imagine a C array that stores count -> bucket for every count up to 1000, and then doing buckets[count_to_bucket[count]] += 1 for another C array buckets, and falling back to something more generic for larger counts (which, naturally, can afford more overhead)).

Not all code is performance critical, but that which runs at every element of every DoFn in every stage in ever pipeline is.

Reply:
Right.  My concern is that there is no amortization done right now, if you look at the corresponding proposed changes in the progress reporting code (i.e. there is no `update_multiple()` implementation on `DistributionCounter`), so there may not be much savings when doing this in a batch.